### PR TITLE
Delete a contact and change its name, closes #11 and closes #12

### LINKED ISSUE #11: Change contact names
- Parent issues: -
- Relevant issues: -
- Estimated effort: 1

* * *

## User story
As a user I want to be able to change the name of a contact so I can always tell who they are.

## What
This issue demands a way to change the name of a contact.

## Why
This issue is needed for the project because the user can add after some time a second contact with the same name and he needs a way to distinguish the two contacts

## How
This feature can be added by providing a button in the contact screen that lets the user change the name of the contact

## Definition of done
This task can be considered done when

- The user can click the button to change the name of a certain contact
- The program asks for the new name of the contact
- The program returns to the contacts list after change the name of the contact
- The name of the contact is changed in the database
- The change in the database is reflected in the contacts list

### LINKED ISSUE #12: Delete contacts
- Parent issues: -
- Relevant issues: Must be done after #73 
- Estimated effort: 1

* * *

## User story
As a user I want to be able to delete someone from my contacts because they're key is no longer valid (or other reasons).

## What
This issue demands a way to delete certain contacts.

## Why
This issue is needed for the project because the key of some contacts can be no longer valid or because the user doesn't want a certain person as contact anymore

## How
This feature can be added by providing a button in the contact screen that lets the user delete the contact

## Definition of done
This task can be considered done when

- The user can click the button to delete a certain contact
- The program asks for confirmation before deleting the contact
- The program returns to the contacts list after deleting the contact
- The contact is deleted in the database
- The change in the database is reflected in the contacts list

### PR DESCRIPTION
**Pull request by clenrock**
- Relevant Issues: #11 #12
- Related Pull Requests: -

* * *

## What
With this PR it's possible to delete a contact and the ability to change the name of the contact.

## Why
Because our app gives users the ability to have a large contact list, it should be possible for them to change the contacts or delete one.

## How
Run the app, login and press a contact. You can change his name by clicking the contact name, typing something and then pressing the save icon. 
When you want to delete a contact you can go to the contact page, and press the bin icon.

## Notes
The changing of the name is not totally intuitive, maybe we should change that a bit.

### CODE REVIEW

File: app/src/androidTest/assets/features/contactActivity.feature [hunk 1]
```diff
@@ -6,7 +6,18 @@ Feature: ContactPage
     When I press the back arrow
     Then I should go to the previous activity I visited
 
-  Scenario: Click on the delete button get popup
+  Scenario: Click on the delete button delete contact
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/contactActivity.feature]: Should be something similar to `Scenario: Deleting a contact`

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java [hunk 2]
```diff
@@ -46,16 +64,21 @@ public ContactActivitySteps(EntryActivity activityClass) {
...
-        IKey key = new SimpleKey("my key", "key");
-        Contact contact = new Contact("Cornel",key );
+        contact = new Contact(TEST_NAME, TEST_KEY);
+
+        serviceLocator = (IServiceLocator) getCurrentActivity().getIntent().getSerializableExtra(ConstantKeywords.SERVICE_LOCATOR);
+        List<Contact> contacts = serviceLocator.getDatabase().getAllContacts();
+        for (Contact contact: contacts) {
+            serviceLocator.getDatabase().deleteContact(contact.getName());
+        }
+        serviceLocator.getDatabase().addContact(contact);
```
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: I think it would be better here to mock the database because you shouldn't (1) rely on the current contents of the database and (2) alter the contents of the database. You can do this by given the `ContactActivity` an intent with a mocked `ServiceLocator`

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java [hunk 3]
```diff
@@ -18,24 +21,39 @@
...
 
 @CucumberOptions(features = "features")
 public class ContactActivitySteps extends ActivityInstrumentationTestCase2<EntryActivity> {
 
-    private Activity parent = null;
+    private Contact contact = null;
+    private IServiceLocator serviceLocator = null;
+    private final static String TEST_NAME = "TestPerson";
+    private final static IKey TEST_KEY = new SimpleKey("my key", "key");
+    private final static String DIFFERENT_NAME = "DifferentName";
```
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: Variables like this one should be set via the `.feature` file, that way you can also easily add test for different values

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 4]
```diff
@@ -174,6 +184,21 @@ public void updateContact(final Contact oldContact, final Contact newContact) th
...
     @Override
+    public Contact getContactWithName(final String contactName) throws IOException {
+        final List<Contact> contacts = getAllContacts();
+        for (final Contact contact: contacts) {
+            if(contact.getName().equals(contactName)) {
+                return contact;
+            }
+        }
+
+        return null;
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Maybe it's better to throw an exception instead of returning null

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 3 above)
Comment (clenrock) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: I don't think that is useful, because we would only have 1 example then. There is no need to have more example than the one described in the code.

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 2 above)
Comment (clenrock) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: When testing, I assume that you start with an empty database so that you have full control over the contents.

**Comment by clenrock** [comment]
@ericcornelissen Good point, I will make the first suggestion

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java [hunk 5]
```diff
@@ -69,13 +92,69 @@ public void iPressDeleteButton() {
...
+    }
+
     @Then("^I should go to the previous activity I visited$")
     public void iShouldGoToPreviousActivity() {
         assertEquals(LoginActivity.class, getCurrentActivity().getClass());
     }
 
-    @Then("^I should get a popup asking if I am sure to delete the contact$")
-    public void iShouldGetPopup() {
+    @Then("^the current contact should be deleted$")
```
Comment (TheBonheurs) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: First letter upper case 🎓

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java [hunk 6]
```diff
@@ -69,13 +92,69 @@ public void iPressDeleteButton() {
...
+
+                    @Override
+                    public void perform(UiController uiController, View view) {
+                        view.performClick();
+                    }
+                }
+        );
+    }
+
+    @Then("^the contact should be updated$")
```
Comment (TheBonheurs) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: See above 👆

File: app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+     * @param v - the view clicked on
+     */
+    public void saveNewContactName(final View v) {
+        final ViewSwitcher switcher = (ViewSwitcher) findViewById(R.id.switch_name_field);
+        final TextView contactName = (TextView) switcher.getNextView();
+        final EditText editText = (EditText) findViewById(R.id.edit_contact_name);
+        switcher.showNext();
+        //Update contact
+        try {
+            final Contact oldContact = serviceLocator.getDatabase().getContactWithName(contactName.getText().toString());
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ChangeContactActivity.java]: Can this part be it's own method see 👇

File: app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java [hunk 8]
```diff
@@ -94,4 +141,66 @@ private void setKeys(final Collection<IKey> keys) {
...
+     * @param v - the view clicked on
+     */
+    public void saveNewContactName(final View v) {
+        final ViewSwitcher switcher = (ViewSwitcher) findViewById(R.id.switch_name_field);
+        final TextView contactName = (TextView) switcher.getNextView();
+        final EditText editText = (EditText) findViewById(R.id.edit_contact_name);
+        switcher.showNext();
+        //Update contact
+        try {
+            final Contact oldContact = serviceLocator.getDatabase().getContactWithName(contactName.getText().toString());
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java]: ⚠️

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 2 above)
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: You shouldn't do that because it messes with the database of the phone/emulator which is not desirable...

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 3 above)
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: Why not? Running tests for multiple different values never hurts...

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 5 above)
Comment (clenrock) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: I don't think this the norm, "Then the current contact should be deleted" is a normal sentence and is wrong when The would start with an uppercase letter

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 3 above)
Comment (clenrock) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: That's true, but it's unnecessary to have multiple strings. I'll still try to change it

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java (hunk 2 above)
Comment (clenrock) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: But for testing purposes it's really useful to control the database like this and to empty the database when running. What do you propose as a solution?

Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ContactActivitySteps.java]: > I think it would be better here to mock the database ... You can do this by given the `ContactActivity` an intent with a mocked `ServiceLocator`

File: app/build.gradle [hunk 9]
```diff
@@ -61,14 +68,11 @@ dependencies {
     compile 'com.android.support:multidex:1.0.1'
     compile 'com.google.zxing:javase:3.3.0'
     compile 'com.google.zxing:core:3.3.0'
-
+    compile 'com.android.support.constraint:constraint-layout:1.0.2'
```
Comment (ericcornelissen) [inline app/build.gradle]: Could you remove this dependency, it is not working for me but everything works fine once I remove it...

Could you also put the `androidTestCompile` back to where it was originally placed 😇

File: app/src/androidTest/assets/features/changeContactActivity.feature [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      And I type a "<differentname>"
+      And I press on the save button
+      Then the contact should be updated
+      And I should go to the previous activity I visited
+
+      Examples:
+        | differentname  |
+        | OtherName  |
+        | 6j206j206j20  |
+        |   |
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/changeContactActivity.feature]: Should this work? I think that an empty (or only spaces) username should not be valid...

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java (hunk 4 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: While making it realized that it is not such a nice method to throw an Exception (Because of all the try/catch statements and the other conflicting Exceptions), so I kept it at returning null.

File: app/src/androidTest/assets/features/changeContactActivity.feature (hunk 10 above)
Comment (clenrock) [inline app/src/androidTest/assets/features/changeContactActivity.feature]: Good point, I will add a check that an empty one should fail

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java [hunk 11]
```diff
@@ -45,6 +45,14 @@
     List<Contact> getAllContacts() throws IOException;
 
     /**
+     * Get one Contact from the database given his name as input.
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java]: His? 🙋

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java [hunk 12]
```diff
@@ -45,6 +45,14 @@
...
+     *
+     * @param contactName - The name of the Contact.
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java]: maybe add that it's a string? very small detail tho

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java (hunk 4 above)
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Maybe it's hadny to put this up as a sticky note or something in this pr/our project as other classes that will depend on this need to be aware of this return strategy.

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 13]
```diff
@@ -84,6 +85,10 @@ private GsonDatabaseAdapter(final IServiceLocator serviceLocator) {
     @Override
     public void addContact(final Contact contact) throws IOException {
         // Get the list of contacts and add the new contact
+        if(getContactWithName(contact.getName()) != null) {
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: I'd appreciate if this was it's own , standradized, method in the db as I had to implement one like this myself but differently.

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 14]
```diff
@@ -114,29 +114,43 @@ private void openContact(final int index) {
...
-        final Contact d = new Contact("Kilian", new SimpleKey("Webmail", "sdjnefiniwfnfejewjnwnkenfk32"));
-        final Contact e = new Contact("Cornel", new SimpleKey("Awesomeness", "nr23uinr3uin2o3uin23oi4un234ijn"));
-        if (!database.getAllContacts().isEmpty()) {
-            database.deleteContact(a);
-            database.deleteContact(b);
-            database.deleteContact(c);
-            database.deleteContact(d);
-            database.deleteContact(e);
+        final IDatabase database = this.serviceLocator.getDatabase();
+        try {
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Why the try catch, I thought Eric said it wasn't needed this morning?

File: app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java [hunk 15]
```diff
@@ -94,4 +88,89 @@ private void setKeys(final Collection<IKey> keys) {
...
+                                            @Override
+                                            public void onClick(final SweetAlertDialog sDialog) {
+                                                sDialog.dismiss();
+                                                finish();
+                                            }
+                                        })
+                                        .showCancelButton(false)
+                                        .changeAlertType(SweetAlertDialog.WARNING_TYPE);
+                            } catch (final IOException e) {
+                                sDialog .setTitleText("Something went wrong")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java]: Add logger?

File: app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java [hunk 16]
```diff
@@ -94,4 +88,89 @@ private void setKeys(final Collection<IKey> keys) {
...
+                                            @Override
+                                            public void onClick(final SweetAlertDialog sDialog) {
+                                                sDialog.dismiss();
+                                                finish();
+                                            }
+                                        })
+                                        .showCancelButton(false)
+                                        .changeAlertType(SweetAlertDialog.SUCCESS_TYPE);
+                            } catch (final IllegalArgumentException e) {
+                                sDialog .setTitleText("The contact doesn't exist")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/ContactActivity.java]: Same as 👇

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 14 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Indeed, the error is catched in the `onCreate` method of this class...

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java (hunk 11 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java]: In English that is a problem, I'll change the sentence.....

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java (hunk 12 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/IDatabase.java]: 👍

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 14 above)
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Good point

File: app/src/androidTest/assets/features/changeContactActivity.feature [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        When I press the contact name
+        And I type "<differentname>"
+        And I close the keyboard
+        And I press on the save button
+        And I press confirm on the popup
+        Then I should stay on the page
+
+        Examples:
+          | differentname  |
+          |   |
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/changeContactActivity.feature]: Since you have only one example, you can make this a `Scenario` instead of `Scenario Outline`

File: app/src/androidTest/assets/features/changeContactActivity.feature [hunk 18]
```diff
@@ -1,43 +0,0 @@
-Feature: ChangeContact
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/changeContactActivity.feature]: Why did you remove this `.feature` file?
